doc/contributing: add lib.optional (#121251)
NixOS/nixpkgs: doc/contributing/coding-conventions.chapter.md

  rev = "${version}";
  ```

> - Filling lists condionally _should_ be done with `lib.optional(s)` instead of using `if cond then [ ... ] else null`.

Reviewer: ```suggestion
- Filling lists condionally _should_ be done with `lib.optional(s)` instead of using `if cond then [ ... ] else []`.
```
This is the implementation of optional(s) and will works with list operators (like `++`).

Author: The null is intentionally to empathies the build avoiding tricks.

Reviewer: I don't get why focusing on the null case.
On master I see 6 occurrences of null versus 177 occurrences of [].

```
pcregrep -rM "]\s+else\s+null" . | wc -l
pcregrep -rM "]\s+else\s+\[\s*\]" . | wc -l
```

Author: Because otherwise the surrounding sentence loses context but we can add both.

```suggestion
- Filling lists condionally _should_ be done with `lib.optional(s)` instead of using `if cond then [ ... ] else null` or `if cond then [ ... ] else [ ]`.
```

  ```nix
  buildInputs = lib.optional stdenv.isDarwin iconv;
  ```

  instead of

  ```nix

>   buildInputs = if stdenv.isDarw then [ iconv ] else null;

Reviewer 2: ```suggestion
  buildInputs = if stdenv.isDarwin then [ iconv ] else null;
```

  ```

>   An exception can be made when fixing a important bug without triggering a mass rebuild.

Author: ```suggestion
  As an exception, an explicit conditional expression with null can be used when fixing a important bug without triggering a mass rebuild.
```

>   If this is done a follow up pull request _should_ be created to change the code to `lib.optional(s)`.

Reviewer: Why would that trigger a rebuild? What do you have in mind?
Furthermore, I guess this is the case for all coding conventions, so I'm not sure that this paragraph is relevant.

Reviewer 3: Re: why — having an `mkDerivation` parameter present but equal to empty list is not the same as not having it at all, and theoretically the builder can observe it (empty vs. unset environment variable).

Reviewer: Indeed, its true when defaulting to `null`, not `[]`.

Author: > Why would that trigger a rebuild? What do you have in mind?

lib.optional returns [] instead of null which triggers a rebuild.

Reviewer: Of course it evaluates to [].

Reviewer 3: Maybe «An exception can be made» should be «The latter can be used»?

I often forget how `null` is handled (and I think it might have changed), so a wording reminding which way actually avoids rebuilds would make it clear why `null` is even mentioned.

Author: Which later? There is no previous sentence which can be referenced, just the example.

Reviewer 3: Well, the example puts two styles into opposition, but you are right that more clarity is better. «As an exception, an explicit conditional expression with `null` can be used when …»?
